src/gorilla.c

```diff
@@ -0,0 +1,261 @@
+  } else {
+    CHECKSPACE(chunk, expectedSize + 1);
+    appendBits(chunk->data, &chunk->idx, 1, 1);
+    appendBits(chunk->data, &chunk->idx, leading, DOUBLE_LEADING);
+    appendBits(chunk->data, &chunk->idx, blockSize - DOUBLE_BLOCK_ADJUST, DOUBLE_BLOCK_SIZE);            
+    appendBits(chunk->data, &chunk->idx, xorWithPrevious >> trailing, blockSize);  
+    chunk->prevLeading = leading;
+    chunk->prevTrailing = trailing;
+  }
+  chunk->prevValue.d = value;
+  return CC_OK;
+}
+
+int CChunk_Append(CompressedChunk *chunk, u_int64_t timestamp, double value) {
+  if (chunk->count == 0) {
+    chunk->baseValue.d   = chunk->prevValue.d = value;
+    chunk->baseTimestamp = chunk->prevTimestamp = timestamp;
+    chunk->prevTimestampDelta = 0;
+  } else {
+    if (appendTS(chunk, timestamp) != CC_OK) return CC_END;
+    if (appendV (chunk, value) != CC_OK) return CC_END;
+  }
+  chunk->count++;
+  return CC_OK;
+}
+
+/********************************** READ *********************************/
+static double readTS(CChunk_Iterator *iter) {
+  int64_t dd = 0;
+  u_int64_t *idx = &iter->idx;
+  u_int64_t *runner = iter->chunk->data;
+
+  // Read stored double delta value
+  if (isBitOff(runner, (*idx)++)) {
+      dd = 0;
+  } else if (isBitOff(runner, (*idx)++)) {
+      dd = bin2int(readBits(runner, idx, 7), 7);
+  } else if (isBitOff(runner, (*idx)++)) {
+      dd = bin2int(readBits(runner, idx, 10), 10);
+  } else if (isBitOff(runner, (*idx)++)) {
+      dd = bin2int(readBits(runner, idx, 13), 13);
+  } else if (isBitOff(runner, (*idx)++)) {
+      dd = bin2int(readBits(runner, idx, 16), 16);
+  } else {
+      dd = bin2int(readBits(runner, idx, 64), 64);
+  }
+
+  // Update iterator
+  iter->prevDelta += dd;
+  return iter->prevTS = iter->prevTS + iter->prevDelta;  
+}
+
+static double readV(CChunk_Iterator *iter) {
+  u_int64_t xorValue;
+  union64bits rv;
+
+  // Check if value was changed
+  if (isBitOff(iter->chunk->data, iter->idx++)) {
+    return iter->prevValue.d;
+  }
+
+  // Check if previous block information was used
+  bool usePreviousBlockInfo = isBitOff(iter->chunk->data, iter->idx++);
+  if (usePreviousBlockInfo) {
+    u_int8_t prevBlockInfo = _64BIT - iter->prevLeading - iter->prevTrailing;
+    xorValue = readBits(iter->chunk->data, &iter->idx, prevBlockInfo);
+    xorValue <<= iter->prevTrailing;
+  } else {
+    u_int64_t leading = readBits(iter->chunk->data, &iter->idx, DOUBLE_LEADING);
+    u_int64_t blocksize = readBits(iter->chunk->data, &iter->idx, DOUBLE_BLOCK_SIZE) + DOUBLE_BLOCK_ADJUST;
+    u_int64_t trailing = _64BIT - leading - blocksize;
+    xorValue = readBits(iter->chunk->data, &iter->idx, blocksize) << trailing;
+    iter->prevLeading = leading;
+    iter->prevTrailing = trailing;    
+  }
+
+  rv.u = xorValue ^ iter->prevValue.u;
+  return iter->prevValue.d = rv.d;
+}
+
+int CChunk_ReadNext(CChunk_Iterator *iter, u_int64_t *timestamp, double *value) {
+  *timestamp = iter->prevTS;
```

Comment:
This can be confusing, there's basically two options here:
1. index == 0 and we should use base TS and base Val
2. index > 0 we should call readTS + readV

The current code suggest you return the previous sample call to read the next one but do not return it.

Reply:
This is being handled at `SeriesIteratorGetNext`